Fix triggering replication when multiple destinations are set

welansari opened the pull request:
- Issue 1: In Artesca "SITE_NAME" is never used, so we always
trigger objects that are replicated to the first storageClass in
the replication rule.

- Issue 2: We check the global replication status when verifying
wether or not an object should be retriggered. This doesn't necessarily
work all the time, especially when replicating to multiple destinations.
As if one destination fails the global status becomes failed, which will
make it impossible to trigger objects with a completed status for example.

- Issue 3: replication info is completely overwritten when it does not contain
info about a specific site. This will cause an issue when replicating to multiple
destinations as the script can only be launched for one site at a time, so when
having a object with non initialized replication info, we won't be able to set
the replication info properly for all destinations.

Issue: S3UTILS-184

bert-e commented:
# Hello kerkesni,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/s3utils/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue S3UTILS-184 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.14.17`

* `1.15.7`


Please check the `Fix Version/s` of S3UTILS-184, or the target
branch of this pull request.

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

Kerkesni commented:
/create_pull_requests

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/1.15/bugfix/S3UTILS-184` with contents from `bugfix/S3UTILS-184`
and `development/1.15`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/1.15/bugfix/S3UTILS-184 origin/development/1.15
 $ git merge origin/bugfix/S3UTILS-184
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/1.15/bugfix/S3UTILS-184
```





*The following options are set:* **create_pull_requests**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/S3UTILS-184` into
`development/1.14`
* [w/1.15/bugfix/S3UTILS-184](https://github.com/scality/s3utils/compare/development/1.15...w/1.15/bugfix/S3UTILS-184) will be merged into `development/1.15` (pull request #332)



The following branches will **NOT** be impacted:

* `development/1.13`
* `development/1.4`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.







*The following options are set:* **create_pull_requests**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_pull_requests**

Commit:
fix triggering replication when multpile destinations are set

- Issue 1: In Artesca "SITE_NAME" is never passed, so we always
trigger objects that are replicated to the first storageClass in
the replication rule.

- Issue 2: We check the global replication status when verifying
wether or not an object should be retriggered. This doesn't necessarily
work all the time, especially when replicating to multiple destinations.
As if one destination fails the global status becomes failed, which will
make it impossible to trigger objects with a completed status for example.

- Issue 3: replication info is completely overwritten when it does not contain
info about a specific site. This will cause an issue when replicating to multiple
destinations as the script can only be launched for one site at a time, so when
having a object with non initialized replication info, we won't be able to set
the replication info propely for all destinations.

Issue: S3UTILS-184

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue S3UTILS-184 contains:


* `1.15.7`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.14.17`

* `1.15.7`


Please check the `Fix Version/s` of S3UTILS-184, or the target
branch of this pull request.




*The following options are set:* **create_pull_requests**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_pull_requests**

francoisferrand commented on CRR/ReplicationStatusUpdater.js:
```diff
@@ -114,16 +114,19 @@ class ReplicationStatusUpdater {
      * Determines if an object should be updated based on its replication metadata properties.
      * @private
      * @param {ObjectMD} objMD - The metadata of the object.
+     * @param {string} storageClass - The storage class for replication.
```
`location` or `site` ?
the `storageClass` is really just where the object is "locally" stored, not where we will copy/send it...

francoisferrand commented on CRR/ReplicationStatusUpdater.js:
```diff
@@ -172,36 +175,54 @@ class ReplicationStatusUpdater {
+                let replicationInfo = objMD.getReplicationInfo();
+                if (!replicationInfo || !replicationInfo.status) {
                     const { Rules, Role } = repConfig;
                     const destination = Rules[0].Destination.Bucket;
```
this may not be correct, if there is more than 1 desitination....

welansari replied:
`crrExistingObjects` doesn't currently support triggering replication for multiple sites at once. So this would work as it will initialize the replication info when triggering the first site and then append the other sites' info to the fields when the other ones are triggered.

francoisferrand replied:
even without triggering multiple sites at once : if a user requests replicatoin to the second "site", this will initialize with the 1st destination (--> trigger replication!), then we will add the second one...

When there is no replicationInfo (i.e. typically when it is empty), should we not just initialize an empty replicationInfo, and let the next block ("Update replication info with site specific info") fill the details for the requested destination?

francoisferrand replied:
in cloudserver, it seems this is initialized to `bucketMD.replicationConfig.destination` : should we do the same?

welansari replied:
This is a weird one... So, we currently only support having a single destination bucket for all replication rules of a bucket. When creating the rules/workflows via UI it's even worse, the destination bucket becomes the name of the bucket we are replicating from.
The value of this field is not used in Zenko. When creating a location, Cloudserver initializes a client class for the respective backend (aws/azure/...) that keeps the name of the destination bucket in memory, that's the value we use when replicating and not what's in the replication rule (only the storageClass is used to know which client to use).

francoisferrand commented on CRR/ReplicationStatusUpdater.js:
```diff
@@ -172,36 +175,54 @@ class ReplicationStatusUpdater {
                         destination,
-                        storageClass,
+                        storageClass: '',
                         role: Role,
-                        storageType: this.storageType,
+                        storageType: '',
```
what are these 2 fields (`storageClass` and `storageType`) used for?

welansari replied:
`storageClass` is what each location's replication queue processor uses to check if it should replicate an object or not. 
`storageType` is used by Cloudserver in Backbeat routes to do some pre-checks (check if versioning is supported on the backend and that the location is valid)

francoisferrand replied:
that is how these fields are used... my question is really what information do they actually store, what these fields represent (for example, the storageClass of the object is already known, and stored in `.location[]` and `dataStoreName` ; and this cannot be the storageClass of the 'remote' object, since there may be multiple destinations...)

same for `destination` btw, I don't understand this field initialize with a single bucket name :-/ Or maybe this is a left-over from the first replication (not supporting multi-targets)?

francoisferrand replied:
the code which initializes these in cloudserver:
```
function _getReplicationInfo(rule, replicationConfig, content, operationType,
    objectMD, bucketMD) {
    const storageTypes = [];
    const backends = [];
    const storageClasses = _getStorageClasses(rule);
    if (!storageClasses) {
        return undefined;
    }
    storageClasses.forEach(storageClass => {
        const storageClassName =
              storageClass.endsWith(':preferred_read') ?
              storageClass.split(':')[0] : storageClass;
        const location = s3config.locationConstraints[storageClassName];
        if (location && replicationBackends[location.type]) {
            storageTypes.push(location.type);
        }
        backends.push(_getBackend(objectMD, storageClassName));
    });
    if (storageTypes.length > 0 && operationType) {
        content.push(operationType);
    }
    return {
        status: 'PENDING',
        backends,
        content,
        destination: replicationConfig.destination,
        storageClass: storageClasses.join(','),
        role: replicationConfig.role,
        storageType: storageTypes.join(','),
        isNFS: bucketMD.isNFS(),
    };
}
```

--> it seems these fields (along with `role`) are not multi-destination/rule aware?
--> these should be mostly useless for Zenko's multibackend replication (otherwise we probably have bugs), but may still be needed for CRR (and may cause bugs if multiple rules/destinations were used in that case)

welansari replied:
`storageClass` and `storageType` are multi-destination aware (although their name suggests otherwise), they contain the list of all storage classes and storage types we are replicating to.
Example:
```
storageClass: "aws-location,azure-blob"
storageType: "aws_s3,azure"
```

In Zenko, these fields are duplicates of information we already have, as backbeat/cloudserver have a list of all location information, we could just use the storage class stored in the rules to get the info we want. I think these are more of a relic from S3C that we can't really remove right now as S3C uses them.

In CRR, the role is also a list. I don't think this works in Zenko tho.
The `destination` field is a weird one, i explained how it works in the previous comment

francoisferrand replied:
> these are more of a relic from S3C that we can't really remove right now as S3C uses them
> In CRR, the role is also a list. I don't think this works in Zenko tho.

"soon", we will have a single arsenal, cloudserver & backbeat : so we can start planning to clean this up, possibly removing (deprecating) redudant fields and aligning the way CRR and multi-backend replication store/manage the state?

can you please create a tech debt ticket, to track this (future) work and document the problem/situation/state of the analysis?

francoisferrand commented on CRR/ReplicationStatusUpdater.js:
```diff
@@ -172,36 +175,54 @@ class ReplicationStatusUpdater {
                     objMD.setReplicationInfo(replicationInfo);
                 }
+                // Update replication info with site specific info
+                if (objMD.getReplicationSiteStatus(storageClass) === undefined) {
```
```suggestion
                if (!objMD.getReplicationSiteStatus(storageClass)) {
```

francoisferrand commented on CRR/ReplicationStatusUpdater.js:
```diff
@@ -114,16 +114,19 @@ class ReplicationStatusUpdater {
                 return (!objMD.getReplicationInfo()
-                    || objMD.getReplicationInfo().status === '');
+                    || !objMD.getReplicationSiteStatus(site));
             }
             return (objMD.getReplicationInfo()
-                && objMD.getReplicationInfo().status === filter);
+                && objMD.getReplicationSiteStatus(site) === filter);
```
is this true as well for CRR (in s3c) ?

esp. we should check that there is no "legacy" data format where we would not have the per-site status, and where we should rely only on the global state...

(e.g. maybe if CRR sometimes has objet which are marked with `replicationInfo.status === 'COMPLETE'` and `site` listed in `storageClass`, but no site status...)

francoisferrand replied:
@nicolas2bert @jonathan-gramain what do you think?

welansari replied:
Replication metadata are managed the same in both S3C and Zenko so yes this should work fine.
I looked back to the initial format of the metadata, there was always a global status.

francoisferrand commented:
code-wise LGTM, but there may be some corner with S3C/CRR, to be confirmed...

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue S3UTILS-184 contains:


* `1.15.7`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.15.8`


Please check the `Fix Version/s` of S3UTILS-184, or the target
branch of this pull request.

Kerkesni commented:
/approve

bert-e commented:
# Build failed



The [build](https://github.com/scality/s3utils/actions?query=branch%3Abugfix/S3UTILS-184) for [commit](https://github.com/scality/s3utils/commit/7df14332ee7071f882783f00e08931753e0a67af) did not succeed in branch [bugfix/S3UTILS-184](https://github.com/scality/s3utils/compare/development/1.15...bugfix/S3UTILS-184)



*The following options are set:* **approve**

Commit:
setup oras in build workflow

oras was removed from from Ubuntu 24.04 GitHub runner images

Issue: S3UTILS-184

bert-e commented:
# Build failed



The [build](https://github.com/scality/s3utils/actions?query=branch%3Abugfix/S3UTILS-184) for [commit](https://github.com/scality/s3utils/commit/4a9f9a1e38c1f5a3d5938642ee62a19ca3269dc1) did not succeed in branch [bugfix/S3UTILS-184](https://github.com/scality/s3utils/compare/development/1.15...bugfix/S3UTILS-184)



*The following options are set:* **approve**

bert-e commented:
# Build failed



The [build](https://github.com/scality/s3utils/actions?query=branch%3Abugfix/S3UTILS-184) for [commit](https://github.com/scality/s3utils/commit/7a943c11919f0b59e823b75f419a93655f73ef75) did not succeed in branch [bugfix/S3UTILS-184](https://github.com/scality/s3utils/compare/development/1.15...bugfix/S3UTILS-184)



*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

Commit:
install libssl1.1 in ci

Ubuntu 24.04 doesn't have libssl1.1 which is needed by mongo-memory-server

Issue: S3UTILS-184

welansari commented:
/approve

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/1.15`



The following branches have **NOT** changed:

* `development/1.13`
* `development/1.14`
* `development/1.4`



Please check the status of the associated issue S3UTILS-184.

Goodbye kerkesni.




*The following options are set:* **approve**
